esp-hal/src/gpio/interconnect.rs

```diff
@@ -1,11 +1,93 @@
-//! Peripheral signal interconnect using IOMUX or GPIOMUX.
-
-// FIXME: https://github.com/esp-rs/esp-hal/issues/2954 The GPIO implementation does not contain any
-// locking. This is okay there, because the implementation uses either W1TS/W1TC
-// registers to set certain bits, or separate registers for each pin - and there
-// can only be one instance of every GPIO struct in safe code. However, with the
-// interconnect module we allow multiple handles, which means possible RMW
-// operations on the pin registers cause data races.
+//! # Peripheral signal interconnect using the GPIO matrix.
```

Comment:
Great documentation! I think it'd be worth adding a sentence or two about IO_MUX. Basically stating that direct connections via IO_MUX means you can't use inversion, filtering, etc.
It's important to know this for the ESP32's SPI at least. If one wants to do 80Mhz transfers.